<a href="https://colab.research.google.com/github/serg-kas/NN-archive/blob/main/%D0%90%D0%BF%D1%80%D0%B5%D0%BB%D1%8C%D1%81%D0%BA%D0%B8%D0%B9_%D0%BA%D1%83%D1%80%D1%81_2020_Light_%D0%94%D0%BE%D0%BC%D0%B0%D1%88%D0%BD%D0%B5%D0%B5_%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_15_%D0%9E%D0%B1%D1%83%D1%87%D0%B0%D1%8E%D1%89%D0%B0%D1%8F_%D0%B8_%D1%82%D0%B5%D1%81%D1%82%D0%BE%D0%B2%D0%B0%D1%8F_%D0%B2%D1%8B%D0%B1%D0%BE%D1%80%D0%BA%D0%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Импорт библиотек

In [ ]:
from tensorflow.keras.models import Sequential # модель прямого распространения
from tensorflow.keras.layers import Dense, Activation, Dropout, BatchNormalization # слои
from tensorflow.keras import utils # тут to_categorical
from tensorflow.keras.preprocessing import image # для отрисовки изображения
from tensorflow.keras.optimizers import Adam, Adadelta # обучение
from tensorflow.keras.datasets import mnist # библиотека с базой MNIST

from sklearn.preprocessing import LabelEncoder, StandardScaler # для нормализации данных
from sklearn import preprocessing # предварительная обработка данных
from sklearn.model_selection import train_test_split  # функция разделения на обучающую и проверочную/тестовую выборку


import numpy as np # массивы numpy
import pandas as pd # pandas

from google.colab import files # Импорт файлов

import matplotlib.pyplot as plt # графики
from PIL import Image # Отрисовка изображений

# Отрисовывать изображения в ноутбуке, а не в консоль или файл
%matplotlib inline


#LIGHT#

Создайте модель для распознавания рукописных цифр из набора MNIST (можно воспользоваться ноутбуком 1-го занятия) и проведите ряд тестов:
  1. Запустите сеть с различными размерами обучающей и проверочной выборок:

      a. Обучающая выборка 50.000 примеров

      b. Обучающая выборка 10.000 примеров

      c. Обучающая выборка 500 примеров
  2. Создайте еще два варианта сети и сравните значения точности на проверочной выборке (на последней эпохе) и на тестовой выборке.Сделайте сравнительную таблицу.
  3. Создайте сеть следующей архитектуры:

    a. 4 Dense слоя

    b. 3 Dropout слоя

    c. 3 BatchNormalization слоя

Напишите свои выводы по результатам проведенных тестов.

In [ ]:
# Повторим сеть для распознавания рукописных цифр из набора MNIST из предыдущего занятия.

# Загружаем MNIST
(x_train_org, y_train_org), (x_test_org, y_test_org) = mnist.load_data()

11493376/11490434 [==============================] - 0s 0us/step


In [ ]:
# Подготовим данные

# Каждый символ из матрицы 28х28 перестроим в вектор 784х1
x_train = x_train_org.reshape(60000,784)
x_test = x_test_org.reshape(10000, 784)

print('Было', x_train_org.shape, 'Стало',  x_train.shape)
print('Было', x_test_org.shape, 'Стало',  x_test.shape)

# Нормализуем входные данные 
#меняем тип на float
x_train = x_train.astype('float32')
#поделив на 255 получим данные в диапазоне от 0 до 1
x_train = x_train / 255 
#меняем тип на float
x_test = x_test.astype('float32')
#поделив на 255 получим данные в диапазоне от 0 до 1
x_test = x_test / 255 

# one_hot_encoding для данных=результатов
y_train = utils.to_categorical(y_train_org, 10)
y_test = utils.to_categorical(y_test_org, 10)

print('Было', y_train_org.shape, 'Стало',  y_train.shape)
print('Было', y_test_org.shape, 'Стало',  y_test.shape)

Было (60000, 28, 28) Стало (60000, 784)
Было (10000, 28, 28) Стало (10000, 784)
Было (60000,) Стало (60000, 10)
Было (10000,) Стало (10000, 10)


In [ ]:
# Создаем сеть (как в предыдущем занятии)
model = Sequential()
#полносвязный слой на 800 нейронов с relu-активацией
model.add(Dense(800, input_dim=784, activation="relu")) 
#полносвязный слой на 400 нейронов с relu-активацией
model.add(Dense(400, activation="relu")) 
#полносвязный слой на 10 нейронов с softmax-активацией
model.add(Dense(10, activation="softmax")) 

# Компиляция
model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"]) 
# Структура модели
print(model.summary())

# Обучение нейронной сети методом fit
model.fit(x_train, y_train, batch_size=128, epochs=15, verbose=1)

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 800)               628000    
_________________________________________________________________
dense_1 (Dense)              (None, 400)               320400    
_________________________________________________________________
dense_2 (Dense)              (None, 10)                4010      
Total params: 952,410
Trainable params: 952,410
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/15
469/469 [==============================] - 1s 3ms/step - loss: 0.2019 - accuracy: 0.9397
Epoch 2/15
469/469 [==============================] - 1s 3ms/step - loss: 0.0761 - accuracy: 0.9762
Epoch 3/15
469/469 [==============================] - 1s 3ms/step - loss: 0.0483 - accuracy: 0.9844
Epoch 4/15
469/469 [==============================] - 1s 3ms/step - loss: 0

In [ ]:
# Вычисляем результаты сети на тестовом наборе
scores = model.evaluate(x_test, y_test, verbose=1)

# scores состоит из двух знанчений
# scores[0] - loss сети на тестовой выборке
# scores[1] - процент правильно распознанных примеров на тестовой выборке
print(scores)
print("Доля верных ответов на тестовых данных, в процентах: ", round(scores[1] * 100, 4), "%", sep="")

313/313 [==============================] - 1s 2ms/step - loss: 0.0868 - accuracy: 0.9833
[0.08679089695215225, 0.983299970626831]
Доля верных ответов на тестовых данных, в процентах: 98.33%


In [ ]:
# Сделаем обучающую выборку 50000 примеров и обучим на ней

# test_size=10000 - будет выделено 10000 примеров данных для тестовой выборки
# shuffle=True - перемешать данные (возможно тут это и не обязательно)
# x_train_50 - данные для обучения
# x_test_50 - данные для проверки
# y_train_50 - правильные ответы для обучения
# y_test_50 - правильные ответы для проверки
x_train_50, x_test_50, y_train_50, y_test_50 = train_test_split(x_train, y_train, test_size=10000, shuffle=True)    
print (x_train_50.shape)
print (x_test_50.shape)
print (y_train_50.shape)
print (y_test_50.shape)

# Создаем сеть (как в предыдущем занятии)
model = Sequential()
#полносвязный слой на 800 нейронов с relu-активацией
model.add(Dense(800, input_dim=784, activation="relu")) 
#полносвязный слой на 400 нейронов с relu-активацией
model.add(Dense(400, activation="relu")) 
#полносвязный слой на 10 нейронов с softmax-активацией
model.add(Dense(10, activation="softmax")) 

# Компиляция
model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"]) 
# Структура модели
print(model.summary())

# Обучение нейронной сети методом fit
model.fit(x_train_50, y_train_50, batch_size=128, epochs=15, verbose=1)


(50000, 784)
(10000, 784)
(50000, 10)
(10000, 10)
Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 800)               628000    
_________________________________________________________________
dense_4 (Dense)              (None, 400)               320400    
_________________________________________________________________
dense_5 (Dense)              (None, 10)                4010      
Total params: 952,410
Trainable params: 952,410
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/15
391/391 [==============================] - 1s 3ms/step - loss: 0.2242 - accuracy: 0.9332
Epoch 2/15
391/391 [==============================] - 1s 3ms/step - loss: 0.0815 - accuracy: 0.9755
Epoch 3/15
391/391 [==============================] - 1s 3ms/step - loss: 0.0524 - accuracy: 0.9828
Epoch 4/15
391/391 [===

In [ ]:
# Вычисляем результаты сети на тестовом наборе
scores = model.evaluate(x_test, y_test, verbose=1)

# scores состоит из двух знанчений
# scores[0] - loss сети на тестовой выборке
# scores[1] - процент правильно распознанных примеров на тестовой выборке
print(scores)
print("Доля верных ответов на тестовых данных, в процентах: ", round(scores[1] * 100, 4), "%", sep="")

313/313 [==============================] - 1s 2ms/step - loss: 0.0873 - accuracy: 0.9815
[0.08730628341436386, 0.9815000295639038]
Доля верных ответов на тестовых данных, в процентах: 98.15%


In [ ]:
# Сделаем обучающую выборку 10000 примеров и обучим на ней

# test_size=50000 - будет выделено 50000 примеров данных для тестовой выборки
# shuffle=True - перемешать данные (возможно тут это и не обязательно)
# x_train_50 - данные для обучения
# x_test_50 - данные для проверки
# y_train_50 - правильные ответы для обучения
# y_test_50 - правильные ответы для проверки
x_train_10, x_test_10, y_train_10, y_test_10 = train_test_split(x_train, y_train, test_size=50000, shuffle=True)    
print (x_train_10.shape)
print (x_test_10.shape)
print (y_train_10.shape)
print (y_test_10.shape)

# Создаем сеть (как в предыдущем занятии)
model = Sequential()
#полносвязный слой на 800 нейронов с relu-активацией
model.add(Dense(800, input_dim=784, activation="relu")) 
#полносвязный слой на 400 нейронов с relu-активацией
model.add(Dense(400, activation="relu")) 
#полносвязный слой на 10 нейронов с softmax-активацией
model.add(Dense(10, activation="softmax")) 

# Компиляция
model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"]) 
# Структура модели
print(model.summary())

# Обучение нейронной сети методом fit
model.fit(x_train_10, y_train_10, batch_size=128, epochs=15, verbose=1)

(10000, 784)
(50000, 784)
(10000, 10)
(50000, 10)
Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 800)               628000    
_________________________________________________________________
dense_7 (Dense)              (None, 400)               320400    
_________________________________________________________________
dense_8 (Dense)              (None, 10)                4010      
Total params: 952,410
Trainable params: 952,410
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/15
79/79 [==============================] - 0s 3ms/step - loss: 0.4988 - accuracy: 0.8503
Epoch 2/15
79/79 [==============================] - 0s 3ms/step - loss: 0.1785 - accuracy: 0.9468
Epoch 3/15
79/79 [==============================] - 0s 3ms/step - loss: 0.1053 - accuracy: 0.9664
Epoch 4/15
79/79 [===========

In [ ]:
# Вычисляем результаты сети на тестовом наборе
scores = model.evaluate(x_test, y_test, verbose=1)

# scores состоит из двух знанчений
# scores[0] - loss сети на тестовой выборке
# scores[1] - процент правильно распознанных примеров на тестовой выборке
print(scores)
print("Доля верных ответов на тестовых данных, в процентах: ", round(scores[1] * 100, 4), "%", sep="")

313/313 [==============================] - 1s 2ms/step - loss: 0.1389 - accuracy: 0.9672
[0.1388876885175705, 0.967199981212616]
Доля верных ответов на тестовых данных, в процентах: 96.72%


In [ ]:
# Сделаем обучающую выборку 500 примеров и обучим на ней

# test_size=59500 - будет выделено 59500 примеров данных для тестовой выборки
# shuffle=True - перемешать данные (возможно тут это и не обязательно)
# x_train_05 - данные для обучения
# x_test_05 - данные для проверки
# y_train_05 - правильные ответы для обучения
# y_test_05 - правильные ответы для проверки
x_train_05, x_test_05, y_train_05, y_test_05 = train_test_split(x_train, y_train, test_size=59500, shuffle=True)    
print (x_train_05.shape)
print (x_test_05.shape)
print (y_train_05.shape)
print (y_test_05.shape)

# Создаем сеть (как в предыдущем занятии)
model = Sequential()
#полносвязный слой на 800 нейронов с relu-активацией
model.add(Dense(800, input_dim=784, activation="relu")) 
#полносвязный слой на 400 нейронов с relu-активацией
model.add(Dense(400, activation="relu")) 
#полносвязный слой на 10 нейронов с softmax-активацией
model.add(Dense(10, activation="softmax")) 

# Компиляция
model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"]) 
# Структура модели
print(model.summary())

# Обучение нейронной сети методом fit
model.fit(x_train_05, y_train_05, batch_size=128, epochs=15, verbose=1)

(500, 784)
(59500, 784)
(500, 10)
(59500, 10)
Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_9 (Dense)              (None, 800)               628000    
_________________________________________________________________
dense_10 (Dense)             (None, 400)               320400    
_________________________________________________________________
dense_11 (Dense)             (None, 10)                4010      
Total params: 952,410
Trainable params: 952,410
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/15
4/4 [==============================] - 0s 3ms/step - loss: 1.9329 - accuracy: 0.4640
Epoch 2/15
4/4 [==============================] - 0s 3ms/step - loss: 0.9437 - accuracy: 0.7940
Epoch 3/15
4/4 [==============================] - 0s 3ms/step - loss: 0.4858 - accuracy: 0.8800
Epoch 4/15
4/4 [=======================

In [ ]:
# Вычисляем результаты сети на тестовом наборе
scores = model.evaluate(x_test, y_test, verbose=1)

# scores состоит из двух знанчений
# scores[0] - loss сети на тестовой выборке
# scores[1] - процент правильно распознанных примеров на тестовой выборке
print(scores)
print("Доля верных ответов на тестовых данных, в процентах: ", round(scores[1] * 100, 4), "%", sep="")

313/313 [==============================] - 1s 2ms/step - loss: 0.4829 - accuracy: 0.8872
[0.48285189270973206, 0.8871999979019165]
Доля верных ответов на тестовых данных, в процентах: 88.72%


[Результаты  обобщены в таблице (см.model)](https://docs.google.com/spreadsheets/d/1SVXnS4zcKhqMv2ZH3apatqHwPedjAsHwOCULNaPk_pU/edit?usp=sharing)

In [ ]:
# Сделаем еще 2 варианта сети протестируем их

# Создаем сеть. Добавим два слоя Dropout.
model_21 = Sequential()
#полносвязный слой на 800 нейронов с relu-активацией
model_21.add(Dense(800, input_dim=784, activation="relu")) 
# Слой Dropout
model_21.add(Dropout(0.3))  
#полносвязный слой на 400 нейронов с relu-активацией
model_21.add(Dense(400, activation="relu")) 
# Слой Dropout
model_21.add(Dropout(0.3))  
#полносвязный слой на 10 нейронов с softmax-активацией
model_21.add(Dense(10, activation="softmax")) 

# Компиляция
model_21.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"]) 
# Структура модели
print(model_21.summary())

# Обучение нейронной сети методом fit
model_21.fit(x_train, y_train, batch_size=128, epochs=15, verbose=1)

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_12 (Dense)             (None, 800)               628000    
_________________________________________________________________
dropout (Dropout)            (None, 800)               0         
_________________________________________________________________
dense_13 (Dense)             (None, 400)               320400    
_________________________________________________________________
dropout_1 (Dropout)          (None, 400)               0         
_________________________________________________________________
dense_14 (Dense)             (None, 10)                4010      
Total params: 952,410
Trainable params: 952,410
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/15
469/469 [==============================] - 1s 3ms/step - loss: 0.2560 - accuracy: 0.9218
Epoch 2/1

In [ ]:
# Вычисляем результаты сети на тестовом наборе
scores = model_21.evaluate(x_test, y_test, verbose=1)

# scores состоит из двух знанчений
# scores[0] - loss сети на тестовой выборке
# scores[1] - процент правильно распознанных примеров на тестовой выборке
print(scores)
print("Доля верных ответов на тестовых данных, в процентах: ", round(scores[1] * 100, 4), "%", sep="")

313/313 [==============================] - 1s 2ms/step - loss: 0.0677 - accuracy: 0.9826
[0.06773972511291504, 0.9825999736785889]
Доля верных ответов на тестовых данных, в процентах: 98.26%


In [ ]:
# Создаем еще вариант. Добавим два слоя BatchNormalization.

model_22 = Sequential()
#полносвязный слой на 800 нейронов с relu-активацией
model_22.add(Dense(800, input_dim=784, activation="relu")) 
# Слой BatchNormalization
model_22.add(BatchNormalization())
#полносвязный слой на 400 нейронов с relu-активацией
model_22.add(Dense(400, activation="relu")) 
# Слой BatchNormalization
model_22.add(BatchNormalization())
#полносвязный слой на 10 нейронов с softmax-активацией
model_22.add(Dense(10, activation="softmax")) 

# Компиляция
model_22.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"]) 
# Структура модели
print(model_22.summary())

# Обучение нейронной сети методом fit
model_22.fit(x_train, y_train, batch_size=128, epochs=15, verbose=1)

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_15 (Dense)             (None, 800)               628000    
_________________________________________________________________
batch_normalization (BatchNo (None, 800)               3200      
_________________________________________________________________
dense_16 (Dense)             (None, 400)               320400    
_________________________________________________________________
batch_normalization_1 (Batch (None, 400)               1600      
_________________________________________________________________
dense_17 (Dense)             (None, 10)                4010      
Total params: 957,210
Trainable params: 954,810
Non-trainable params: 2,400
_________________________________________________________________
None
Epoch 1/15
469/469 [==============================] - 2s 4ms/step - loss: 0.1653 - accuracy: 0.9493
Epoch

In [ ]:
# Вычисляем результаты сети на тестовом наборе
scores = model_22.evaluate(x_test, y_test, verbose=1)

# scores состоит из двух знанчений
# scores[0] - loss сети на тестовой выборке
# scores[1] - процент правильно распознанных примеров на тестовой выборке
print(scores)
print("Доля верных ответов на тестовых данных, в процентах: ", round(scores[1] * 100, 4), "%", sep="")

313/313 [==============================] - 1s 2ms/step - loss: 0.0964 - accuracy: 0.9782
[0.0963997095823288, 0.9782000184059143]
Доля верных ответов на тестовых данных, в процентах: 97.82%


[Результаты  обобщены в таблице (см.model_21 и model_22)](https://docs.google.com/spreadsheets/d/1SVXnS4zcKhqMv2ZH3apatqHwPedjAsHwOCULNaPk_pU/edit?usp=sharing)

In [ ]:
# Создадим сеть следующей архитектуры: 4 Dense слоя, 3 Dropout слоя 3 BatchNormalization слоя

model_3 = Sequential()
#полносвязный слой на 800 нейронов с relu-активацией
model_3.add(Dense(800, input_dim=784, activation="relu")) 
# Слой BatchNormalization
model_3.add(BatchNormalization())
# Слой Dropout
model_3.add(Dropout(0.3))
#полносвязный слой на 400 нейронов с relu-активацией
model_3.add(Dense(400, activation="relu")) 
# Слой BatchNormalization
model_3.add(BatchNormalization())
# Слой Dropout
model_3.add(Dropout(0.3))
#полносвязный слой на 800 нейронов с relu-активацией
model_3.add(Dense(800, input_dim=784, activation="relu")) 
# Слой BatchNormalization
model_3.add(BatchNormalization())
# Слой Dropout
model_3.add(Dropout(0.3))
#полносвязный слой на 400 нейронов с relu-активацией
model_3.add(Dense(400, activation="relu")) 
#полносвязный слой на 10 нейронов с softmax-активацией
model_3.add(Dense(10, activation="softmax")) 

# Компиляция
model_3.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"]) 
# Структура модели
print(model_3.summary())

# Обучение нейронной сети методом fit
model_3.fit(x_train, y_train, batch_size=128, epochs=15, verbose=1)

Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_18 (Dense)             (None, 800)               628000    
_________________________________________________________________
batch_normalization_2 (Batch (None, 800)               3200      
_________________________________________________________________
dropout_2 (Dropout)          (None, 800)               0         
_________________________________________________________________
dense_19 (Dense)             (None, 400)               320400    
_________________________________________________________________
batch_normalization_3 (Batch (None, 400)               1600      
_________________________________________________________________
dropout_3 (Dropout)          (None, 400)               0         
_________________________________________________________________
dense_20 (Dense)             (None, 800)              

In [ ]:
# Вычисляем результаты сети на тестовом наборе
scores = model_3.evaluate(x_test, y_test, verbose=1)

# scores состоит из двух знанчений
# scores[0] - loss сети на тестовой выборке
# scores[1] - процент правильно распознанных примеров на тестовой выборке
print(scores)
print("Доля верных ответов на тестовых данных, в процентах: ", round(scores[1] * 100, 4), "%", sep="")

313/313 [==============================] - 1s 2ms/step - loss: 0.0624 - accuracy: 0.9838
[0.06240004301071167, 0.9837999939918518]
Доля верных ответов на тестовых данных, в процентах: 98.38%


[Результаты  обобщены в таблице (см.model_3)](https://docs.google.com/spreadsheets/d/1SVXnS4zcKhqMv2ZH3apatqHwPedjAsHwOCULNaPk_pU/edit?usp=sharing)

#PRO#

*Вариант 1*


Повысьте точность модели по обнаружению мин до 90 % на тестовой выборке. Можно использовать
различные варианты слоев Dropout и BatchNormalization. Можно менять количество примеров в
обучающей и проверочной выборках, но нельзя менять количество примеров в тестовой.



In [ ]:
# Повторим модель по обнаружению  мин из вебинара

# Загружаем файл sonar
files.upload()
!ls # Выводим содержимое корневой папки

Saving sonar.csv to sonar.csv
sample_data  sonar.csv


In [ ]:
# Загружаем данные из файла sonar.csv
df = pd.read_csv("sonar.csv", header=None)  
df.head()  

0       1       2       3       4   ...      56      57      58      59  60
0  0.0200  0.0371  0.0428  0.0207  0.0954  ...  0.0180  0.0084  0.0090  0.0032   R
1  0.0453  0.0523  0.0843  0.0689  0.1183  ...  0.0140  0.0049  0.0052  0.0044   R
2  0.0262  0.0582  0.1099  0.1083  0.0974  ...  0.0316  0.0164  0.0095  0.0078   R
3  0.0100  0.0171  0.0623  0.0205  0.0205  ...  0.0050  0.0044  0.0040  0.0117   R
4  0.0762  0.0666  0.0481  0.0394  0.0590  ...  0.0072  0.0048  0.0107  0.0094   R

[5 rows x 61 columns]

In [ ]:
print(df.shape) # Размерность исходных данных

(208, 61)


In [ ]:
dataset = df.values                 # Берем только значения массива(без индексов)
X = dataset[:,0:60].astype(float)   # Присваиваем им тип данных - float(с плавающей точкой) данным с 0 по 60 колонки
Y = dataset[:,60]                   # Присваеваем значению Y данные из столбца с индексом 60
Y[Y=='R']='0'                       # Если значение элемента столбца равно 'R', присваеваем ему значение '0'
Y[Y=='M']='1'                       # Если значение элемента столбца равно 'Y', присваеваем ему значение '1'
Y = Y.astype(int)                   # Меняем тип данных столбца на 'int'(целочисленный тип данных)
print(X.shape)                      # Выводим размерность X
print(Y.shape)                      # Выводим размерность Y
print(Y)                            

(208, 60)
(208,)
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]


In [ ]:
#Выделим тестовую выборку

# test_size=0.2 - будет выделено 20% от тренировочных данных 
# shuffle=True - перемешать данные
# x_train - данные для обучения
# x_test - данные для проверки
# y_train - правильные ответы для обучения
# y_test - правильные ответы для проверки
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, shuffle=True)    
print (x_train.shape)
print (x_test.shape)
print (y_train.shape)
print (y_test.shape)

(166, 60)
(42, 60)
(166,)
(42,)


In [ ]:
# Создаем сеть
model = Sequential()

# Добавляем слои
model.add(Dense(60, input_dim=60, activation='relu'))
model.add(Dense(30,  activation='relu'))
model.add(Dense(1, activation='sigmoid'))

# Компилируем сеть
model.compile(loss='binary_crossentropy', optimizer=Adam(lr=0.001), metrics=['accuracy'])

# Обучаем сеть
model.fit(x_train,         
          y_train,         
          batch_size=8,    # Размер батча
          epochs=100,      # Количество эпох
          verbose=1)       # Выводить процесс обучения на каждой эпохе

          

Epoch 1/100
21/21 [==============================] - 0s 3ms/step - loss: 0.6940 - accuracy: 0.5181
Epoch 2/100
21/21 [==============================] - 0s 3ms/step - loss: 0.6565 - accuracy: 0.6867
Epoch 3/100
21/21 [==============================] - 0s 3ms/step - loss: 0.6347 - accuracy: 0.6867
Epoch 4/100
21/21 [==============================] - 0s 3ms/step - loss: 0.6164 - accuracy: 0.6747
Epoch 5/100
21/21 [==============================] - 0s 3ms/step - loss: 0.5867 - accuracy: 0.7229
Epoch 6/100
21/21 [==============================] - 0s 3ms/step - loss: 0.5597 - accuracy: 0.7771
Epoch 7/100
21/21 [==============================] - 0s 3ms/step - loss: 0.5416 - accuracy: 0.7530
Epoch 8/100
21/21 [==============================] - 0s 4ms/step - loss: 0.5077 - accuracy: 0.7771
Epoch 9/100
21/21 [==============================] - 0s 3ms/step - loss: 0.4759 - accuracy: 0.8133
Epoch 10/100
21/21 [==============================] - 0s 3ms/step - loss: 0.4549 - accuracy: 0.8012
Epoch 11/

In [ ]:
# Вычисляем результаты сети на тестовом наборе
scores = model.evaluate(x_test, y_test, verbose=1)

# scores состоит из двух знанчений
# scores[0] - loss сети на тестовой выборке
# scores[1] - процент правильно распознанных примеров на тестовой выборке
print(scores)
print("Доля верных ответов на тестовых данных, в процентах: ", round(scores[1] * 100, 4), "%", sep="")

2/2 [==============================] - 0s 4ms/step - loss: 0.7758 - accuracy: 0.8333
[0.7758095264434814, 0.8333333134651184]
Доля верных ответов на тестовых данных, в процентах: 83.3333%


Итак, модель из вебинара обучается на 1, но на тестовой выборке дает точность 76%. Будем стараться поднять ее до 90% путем изменения архитектуры сети. 

По условию  задачи будем использовать тестовую выборку 42 элемента как в вебинаре. 

Обучающую выборку не будем делить еще на обучающую и валидационную. Так как ее размер 166 элементов и так весьма не велик.

In [ ]:
# В первую очередь пробуем добавлять слои Dropout и BatchNormalization 
model = Sequential()

model.add(BatchNormalization(input_shape=(60,)))
#model.add(Dropout(0.3, input_shape=(60,))) 

model.add(Dropout(0.1))

model.add(Dense(60, activation='relu'))

#model.add(Dropout(0.2))

model.add(BatchNormalization())

model.add(Dense(30,  activation='relu'))

model.add(Dropout(0.2))

model.add(Dense(1, activation='sigmoid'))

# Компилируем сеть
model.compile(loss='binary_crossentropy', optimizer=Adam(lr=0.001), metrics=['accuracy'])

# Структура модели
print(model.summary())

# Обучаем сеть
model.fit(x_train,         
          y_train,         
          batch_size=8,    # Размер батча
          epochs=100,      # Количество эпох
          verbose=1)       # Выводить процесс обучения на каждой эпохе


Model: "sequential_114"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
batch_normalization_211 (Bat (None, 60)                240       
_________________________________________________________________
dropout_241 (Dropout)        (None, 60)                0         
_________________________________________________________________
dense_346 (Dense)            (None, 60)                3660      
_________________________________________________________________
batch_normalization_212 (Bat (None, 60)                240       
_________________________________________________________________
dense_347 (Dense)            (None, 30)                1830      
_________________________________________________________________
dropout_242 (Dropout)        (None, 30)                0         
_________________________________________________________________
dense_348 (Dense)            (None, 1)              

In [ ]:
# Вычисляем результаты сети на тестовом наборе
scores = model.evaluate(x_test, y_test, verbose=1)

# scores состоит из двух знанчений
# scores[0] - loss сети на тестовой выборке
# scores[1] - процент правильно распознанных примеров на тестовой выборке
print(scores)
print("Доля верных ответов на тестовых данных, в процентах: ", round(scores[1] * 100, 4), "%", sep="")

2/2 [==============================] - 0s 4ms/step - loss: 0.2528 - accuracy: 0.9286
[0.2527619004249573, 0.9285714030265808]
Доля верных ответов на тестовых данных, в процентах: 92.8571%


Искомой точности на тестовой выборке кое-как добились, но провалы по точности все таки бывают.